paymentsheet/src/main/java/com/stripe/android/paymentelement/embedded/form/PrimaryButtonStateHolder.kt

```diff
@@ -0,0 +1,71 @@
+package com.stripe.android.paymentelement.embedded.form
+
+import com.stripe.android.core.injection.ViewModelScope
+import com.stripe.android.core.strings.ResolvableString
+import com.stripe.android.lpmfoundations.paymentmethod.PaymentMethodMetadata
+import com.stripe.android.model.PaymentIntent
+import com.stripe.android.model.StripeIntent
+import com.stripe.android.paymentelement.EmbeddedPaymentElement
+import com.stripe.android.paymentelement.ExperimentalEmbeddedPaymentElementApi
+import com.stripe.android.paymentelement.embedded.EmbeddedSelectionHolder
+import com.stripe.android.paymentsheet.model.amount
+import com.stripe.android.paymentsheet.model.currency
+import com.stripe.android.paymentsheet.ui.PrimaryButtonProcessingState
+import com.stripe.android.paymentsheet.utils.buyButtonLabel
+import com.stripe.android.ui.core.Amount
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.collectLatest
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@OptIn(ExperimentalEmbeddedPaymentElementApi::class)
+internal class PrimaryButtonStateHolder @Inject constructor(
+    paymentMethodMetadata: PaymentMethodMetadata,
+    selectionHolder: EmbeddedSelectionHolder,
+    @ViewModelScope coroutineScope: CoroutineScope,
+    configuration: EmbeddedPaymentElement.Configuration,
+) {
+    private val _state = MutableStateFlow(
+        State(
+            label = primaryButtonLabel(paymentMethodMetadata.stripeIntent, configuration),
+            isEnabled = false,
+            processingState = PrimaryButtonProcessingState.Idle(null)
+        )
+    )
+    val state: StateFlow<State> = _state
+
+    init {
+        coroutineScope.launch {
+            selectionHolder.selection.collectLatest { selection ->
+                _state.update { state ->
+                    state.copy(
+                        isEnabled = selection != null
+                    )
+                }
+            }
+        }
+    }
+
+    private fun primaryButtonLabel(
+        stripeIntent: StripeIntent,
+        configuration: EmbeddedPaymentElement.Configuration
```

Comment:
Looks like you shouldn't need to pass these, since they're injected. You can just read the properties.

Probably just personal preference though, so take it or leave it.